### ISSUE TITLE
End of Math block sometimes cut off or accelerated in speech

### ISSUE THREAD

**Issue by bhavyashah** [NSoiffer/MathCAT#187]
I was reproducing this fairly consistently earlier today and now it is a 50-50. In the same NVDA session (no restarting NVDA, no file reopening, no disabling/re-enabling MathCAT add-on), I can experience this bug sometimes and not other times (both seem roughly equally likely). What this looks like for a simple Math block that reads "lambda sub 2" is as follows:
Normal: lambda sub 2, slow and steady and respecting the relative speech rate and everything is spoken normally
Buggy: lambda 2, with the sub entirely inaudible and the "oo" in the 2 very sped up or cut off
This is making it harder to process small Math blocks (that contain just a sigma or X or such) as well as the endings of Math blocks (esp. if they contain an important number). There is no difference in the NVDA log for both these behaviours. I am using Microsoft Edge 117.0.2045.47, NVDA 2023.3beta3, and MathCAT version 0.3.6. 
I arrow up to the Math block and it may read it in a normal or buggy manner, but identically log as follows:
IO - speech.speech.speak (22:50:31.219) - MainThread (9220):
Speaking ['clickable', RateCommand(multiplier=0.56), ' ', 'lambda sub 2', ' ', RateCommand()]

**Comment by NSoiffer** [comment]
What speech engine are you using? espeak? SAPI? OneCore?

**Comment by bhavyashah** [comment]
ESpeak-NG, Voice: English (Great Britain), Variant: Steph3.

**Comment by andrew-l-d** [comment]
That is what I am using too, so that is not the issue.

**Comment by bhavyashah** [comment]
Confirming that this issue persists post a restart.
I am trying to test MathCAT with all other NVDA add-ons disabled to ensure that is not the source of this issue. Will update this ticket once that testing is done.
While I only spent a few minutes with SAPI and OneCore, this issue may be specific to ESpeak-NG. For Math which ESpeak-NG sped up the end of, OneCore and SAPI enunciated fully and at a steady rate. It is possible though that if I spent enough time with SAPI or OneCore, I might be able to reproduce this issue with them too. For now, let's say this issue occurs only when ESpeak-NG is in use.

**Comment by NSoiffer** [comment]
@bhavyashah: Can you put the MathML that is (sometimes) causes problems into the issue? That will hopefully help me figure out what is going on.

**Comment by bhavyashah** [comment]
Example: $k=2,3,4$
Use ESpeak-NG set to Rate 35 with Rate boost enabled and MathCAT's Relative speech rate at 25. I was able to reproduce this on the above example near-consistently with ESpeak NG at 35 with rate boost enabled, but that ceased when I brought down the speech rate to 0 with rate boost enabled or switched to OneCore. Oddly specific configuration.

**Comment by bhavyashah** [comment]
This issue likely has something to do with speech rates and ESpeak-NG. I have been reproducing this pretty consistently with short Math blocks the last few days where the end of a Math block is distinctly accelerated or garbled. When I reduce speech rate though, at some point the issue gets fixed. Earlier today, moving from 35 to 30 fixed it. Right now, I had to go from 40 to 15 to fix the issue. Quite strange.
There is another issue where for long Math blocks, NVDA decelerates dramatically midway through the Math and then speaks increasingly slowly. Testing to determine whether that too is exclusive to ESpeak-NG, a test case, and a separate ticket are necessary for that. I do wonder though if that issue is related to the one reported in this ticket.

**Comment by NSoiffer** [comment]
This looks like it is going to be hard to debug...

I haven't been able to reproduce the slowing, but I did get the "4" cut off at the end problem (eventually). I'm not exactly using the same relative speech rate because the current version of the dialog (rhs=100% vs middle being 100% in older dialog) doesn't get me to a relative value of 25 (the closest is 22).

If you come up with any other ideas that make it reproducible, let me know.

The following error showed up in the log message, but it stopped showing up. It didn't reappear if I reloaded the page:
```
ERROR - scriptHandler.executeScript (14:42:06.080) - MainThread (46224):
error executing script: <bound method CursorManager.script_moveByLine_forward of <NVDAObjects.IAccessible.chromium.ChromeVBuf object at 0x089CDE70>> with gesture 'down arrow'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 295, in executeScript
  File "cursorManager.pyc", line 274, in script_moveByLine_forward
  File "cursorManager.pyc", line 171, in _caretMovementScriptHelper
  File "speech\speech.pyc", line 1231, in speakTextInfo
  File "speech\types.pyc", line 42, in __iter__
  File "speech\speech.pyc", line 1402, in getTextInfoSpeech
  File "speech\speech.pyc", line 1203, in _extendSpeechSequence_addMathForTextInfo
  File "virtualBuffers\__init__.pyc", line 406, in getMathMl
AttributeError: 'NoneType' object has no attribute 'mathMl'
```

I did note one other strangeness. The "4" was cut off only on the down arrow, not on the up arrow. However, after reloading the page, it was cut off only on the up arrow, not the down arrow.

In the above message, none of the MathCAT files are mentioned. But it does mention a virtual buffer. So I suspect for speech getting cut off, it is a problem with math and the virtual buffer.

(mainly for my benefit)
The error in the log happens on the last line of
```
	def getMathMl(self, field):
		docHandle = int(field["controlIdentifier_docHandle"])
		nodeId = int(field["controlIdentifier_ID"])
		obj = self.obj.getNVDAObjectFromIdentifier(docHandle, nodeId)
		return obj.mathMl
```
Hence, `getNVDAObjectFromIdentifier` is return None.

I believe the code for that is the version in MSHTML.py:
```
	def getNVDAObjectFromIdentifier(self, docHandle, ID):
		HTMLNode=NVDAObjects.IAccessible.MSHTML.locateHTMLElementByID(self.rootNVDAObject.HTMLNode.document,'ms__id%d'%ID)
		if not HTMLNode:
			return self.rootNVDAObject
		return NVDAObjects.IAccessible.MSHTML.MSHTML(HTMLNode=HTMLNode)
```
